modules/services/gromit-mpx.nix

```diff
@@ -0,0 +1,212 @@
+  undokey =
+    if cfg.undoKeyCode != null
+      then "--undo-keycode ${toString cfg.undoKeyCode}"
+      else if cfg.undoKey != null
+        then "--undo-key ${cfg.undoKey}"
+        else "--undo-key none";
+
+  # The command line to send to gromit-mpx:
+  commandArgs = concatStringsSep " " [
+    hotkey
+    undokey
+    "--opacity ${toString cfg.opacity}"
+  ];
+
+  # Allowed modifiers:
+  modsAndButtons = [
+    "1" "2" "3" "4" "5"
+    "SHIFT" "CONTROL" "ALT" "META"
+  ];
+
+  # Create a string of tool attributes:
+  toolAttrs = tool: concatStringsSep " " (
+    [ "size=${toString tool.size}" ] ++
+    optional (tool.type != "eraser") ''color="${tool.color}"'' ++
+    optional (tool.arrowSize != null) "arrowsize=${toString tool.arrowSize}");
+
+  # Optional tool modifier string:
+  toolMod = tool:
+    if tool.modifiers != []
+      then "[" + concatStringsSep ", " tool.modifiers + "]"
+      else "";
+
+  # A single tool configuration:
+  toolToCfg = n: tool: ''
+    "tool-${toString n}" = ${toUpper tool.type} (${toolAttrs tool});
+    "${tool.device}"${toolMod tool} = "tool-${toString n}";
+  '';
+
+  # Per-tool options:
+  toolOptions = {
+    options = {
+      device = mkOption {
+        type = types.str;
+        example = "default";
+        description = ''
+          Use this tool with the given xinput device.  The device with
+          the name default works with any input.
+        '';
+      };
+
+      type = mkOption {
+        type = types.enum [ "pen" "eraser" "recolor" ];
+        default = "pen";
+        example = "eraser";
+        description = "Which type of tool this is.";
+      };
+
+      color = mkOption {
+        type = types.str;
+        default = "red";
+        example = "#ff00ff";
+        description = "The stroke (or recolor) color of the tool.";
+      };
+
+      size = mkOption {
+        type = types.ints.positive;
+        default = 5;
+        example = 3;
+        description = "The tool size.";
+      };
+
+      arrowSize = mkOption {
+        type = types.nullOr types.ints.positive;
+        default = null;
+        example = 2;
+        description = ''
+          If not null, automatically draw an arrow at the end of a
+          stroke with the given size.
+        '';
+      };
+
+      modifiers = mkOption {
+        type = types.listOf (types.enum modsAndButtons);
+        default = [];
+        example = [ "SHIFT" ];
+        description = ''
+          Only activate this tool if the given modifiers are also active.
+        '';
+      };
+    };
+  };
+
+in
+{
+  meta.maintainers = [ maintainers.pjones ];
+
+  options.services.gromit-mpx = {
+    enable = mkEnableOption "Gromit-MPX annotation tool";
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.gromit-mpx;
+      defaultText = "pkgs.gromit-mpx";
+      description = "The gromit-mpx package to use.";
+    };
+
+    hotKey = mkOption {
+      type = types.nullOr types.str;
```

Comment:
Could this have the type
```nix
with types; nullOr (either str ints.positive)
```
where a string is interpreted as an keysym and an integer as a key code? Would be nice if so since then the `hotKeyCode` option could be removed.

Similar holds for the `undoKey` option.